mesh/ncmesh.hpp

```diff
@@ -375,7 +377,7 @@ class NCMesh
        to its vertex nodes. */
    struct Element
    {
-      char geom;     ///< Geometry::Type of the element
+      Geometry::Type geom; ///< Geometry::Type of the element
```

Comment:
Can we make sure the storage type of Geometry::Type is char? (Not sure this can be done in C++03). The char was intentional here to save memory (it will get packed with the other chars here).

Reply:
I think an `enum` type is much more appropriate here, I would be curious to see if this "optimization" results in any computation time difference.

Reply:
Of course it's appropriate. I'm just suggesting it's storage type is char. The difference is not negligible since everything is memory-bound today so having a smaller data structure is better.

Reply:
I am in favor of smaller enumerated types but I'm fairly certain you cannot specify the underlying type before c++11.  Perhaps we can use the preprocessor to set this type if the compiler is c++11 compliant.

Reply:
Actually, I think the size (on my laptop at least) goes from 52 bytes to 56 which is less than 8%.  Not trivial but not 35% either.

Reply:
We could have an #ifdef for C++11 in geom.hpp to do `enum Type : char {...}` but since it's only 8% I'm willing to give it up :-)

Reply:
I'm confused now, I thought a `char` was 1 byte, an `int` 4 bytes... So it should be 11 bytes vs 8 bytes no? Am I missing something?? Is it due to some memory alignment? (11/8=1.375)

Reply:
I have 12 bytes vs 8 bytes on my desktop... so that's 50%

Reply:
I'll check to see if there's a standard variable we could use for this #ifdef and put it in if I find one.

Reply:
@YohannDudouit , how could it be 8 bytes?  The `NCMesh::Element` struct contains three `char`s, 4 `int`s, and a `union` the size of 8 `int`s.  We're talking about 3 `char`s and 12 `int`s or 2 `char`s and 13 `int`s.  Perhaps you're looking at a different object?

Reply:
@mlstowell Oh sorry... Git squished:
```
int rank;      ///< processor number (ParNCMesh), -1 if undefined/unknown
       int attribute;
       union
       {
          int node[8];  ///< element corners (if ref_type == 0)
int child[8]; ///< 2-8 children (if ref_type != 0)
```